internal/tofu/eval_context_builtin.go

```diff
@@ -408,6 +411,108 @@ func (ctx *BuiltinEvalContext) EvaluateReplaceTriggeredBy(expr hcl.Expression, r
 	return ref, replace, diags
 }
 
+// EvaluateImportAddress takes the raw reference expression of the import address
+// from the config, and returns the evaluated address addrs.AbsResourceInstance
+//
+// The implementation is inspired by config.AbsTraversalForImportToExpr, but this time we can evaluate the expression
+// in the indexes of expressions. If we encounter a hclsyntax.IndexExpr, we can evaluate the Key expression and create
+// an Index Traversal, adding it to the Traverser
+func (ctx *BuiltinEvalContext) EvaluateImportAddress(expr hcl.Expression) (addrs.AbsResourceInstance, tfdiags.Diagnostics) {
+	traversal, diags := ctx.traversalForImportExpr(expr)
+	if diags.HasErrors() {
+		return addrs.AbsResourceInstance{}, diags
+	}
+
+	return addrs.ParseAbsResourceInstance(traversal)
+}
+
+func (ctx *BuiltinEvalContext) traversalForImportExpr(expr hcl.Expression) (traversal hcl.Traversal, diags tfdiags.Diagnostics) {
+	switch e := expr.(type) {
+	case *hclsyntax.IndexExpr:
+		t, d := ctx.traversalForImportExpr(e.Collection)
+		diags = diags.Append(d)
+		traversal = append(traversal, t...)
+
+		tIndex, dIndex := ctx.parseImportIndexKeyExpr(e.Key)
+		diags = diags.Append(dIndex)
+		traversal = append(traversal, tIndex)
+	case *hclsyntax.RelativeTraversalExpr:
+		t, d := ctx.traversalForImportExpr(e.Source)
+		diags = diags.Append(d)
+		traversal = append(traversal, t...)
+		traversal = append(traversal, e.Traversal...)
+	case *hclsyntax.ScopeTraversalExpr:
+		traversal = append(traversal, e.Traversal...)
+	default:
+		// This should not happen, as it should have failed validation earlier, in config.AbsTraversalForImportToExpr
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid import address expression",
+			Detail:   "Import address must be a reference to a resource's address, and only allows for indexing with dynamic keys. For example: module.my_module[expression1].aws_s3_bucket.my_buckets[expression2] for resources inside of modules, or simply aws_s3_bucket.my_bucket for a resource in the root module",
+			Subject:  expr.Range().Ptr(),
+		})
+	}
+	return
+}
+
+// parseImportIndexKeyExpr parses an expression that is used as a key in an index, of an HCL expression representing an
+// import target address, into a traversal of type hcl.TraverseIndex.
+// After evaluation, the expression must be known, not null, not sensitive, and must be a string (for_each) or a number
+// (count)
+func (ctx *BuiltinEvalContext) parseImportIndexKeyExpr(expr hcl.Expression) (hcl.TraverseIndex, tfdiags.Diagnostics) {
+	idx := hcl.TraverseIndex{
+		SrcRange: expr.Range(),
+	}
+
+	val, diags := ctx.EvaluateExpr(expr, cty.DynamicPseudoType, nil)
+	if diags.HasErrors() {
+		return idx, diags
+	}
+
+	if !val.IsKnown() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Import block 'to' address contains invalid key",
```

Comment:
```suggestion
			Summary:  "Import block 'to' address contains an invalid key",
```

Reply:
Thanks for catching these https://github.com/opentofu/opentofu/pull/1270/commits/ac25cf3ebf6457893dfc4412ff594d38093042a6